lib/auth/autoupdate_rollout.go

```diff
@@ -0,0 +1,174 @@
+		if !isCatchAll {
+			// No need to check for UpdaterInfo being nil, it would have been filtered
+			// out by filterHandler().
+			return handle.Hello().UpdaterInfo.UpdateGroup == groupName
+		}
+		// This is the catch-call group, it matches agents from every group not in groups.
+		_, ok = groupSet[groupName]
+		// If the agent group is not in the group list, it falls into the catch-all.
+		return !ok
+	}
+
+	sampler := newHandlerSampler(sampleSize, filter)
+
+	a.inventory.UniqueHandles(sampler.visit)
+
+	sampled := sampler.Sampled()
+	canaries := make([]*autoupdatev1pb.Canary, len(sampled))
+	for i, h := range sampled {
+		hello := h.Hello()
+		updaterID, err := uuid.FromBytes(hello.UpdaterInfo.UpdateUUID)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		canaries[i] = &autoupdatev1pb.Canary{
+			UpdaterId: updaterID.String(),
+			HostId:    hello.ServerID,
+			Hostname:  hello.Hostname,
+			Success:   false,
+		}
+	}
+	return canaries, nil
+}
+
+// handleSampler randomly samples handles from the inventory.
+// It implements Alan Waterman's Reservoir Sampling Algorithm R
+// (The Art of Computer Programming Volume 2).
+// See https://en.wikipedia.org/wiki/Reservoir_sampling for more details.
+type handleSampler struct {
+	sampleSize int
+	seenCount  int
+	filter     func(handle inventory.UpstreamHandle) bool
+	// TODO for reviewers:
+	// Do we feel confident about holding to the Handle even after we're done visiting?
+	// I think so but @espadolini had doubts.
+	// Alternatives are:
+	// - using generics and taking a func(inventory.UpstreamHandle) (K, bool)
+	// - making the sampler part of the inventory package
+	sample []inventory.UpstreamHandle
+}
+
+func newHandlerSampler(sampleSize int, filter func(handle inventory.UpstreamHandle) bool) *handleSampler {
+	return &handleSampler{
+		sampleSize: sampleSize,
+		seenCount:  sampleSize,
+		filter:     filter,
+		sample:     make([]inventory.UpstreamHandle, 0, sampleSize),
+	}
+}
+
+// pass this function to inventory.AllHandles()
+func (h *handleSampler) visit(handle inventory.UpstreamHandle) {
+	// filter out everything we don't want
+	if !h.filter(handle) {
+		return
+	}
+
+	// Fill the reservoir
+	if len(h.sample) < h.sampleSize {
+		h.sample = append(h.sample, handle)
+		h.seenCount++
+		return
+	}
+
+	// Reservoir is already filled, replace existing elements.
+	if j := rand.N(h.seenCount); j < h.sampleSize {
+		h.sample[j] = handle
+	}
+	h.seenCount++
+}
+
+func (h *handleSampler) Sampled() []inventory.UpstreamHandle {
+	return h.sample
+}
+
+func (a *Server) LookupAgentInInventory(ctx context.Context, hostID string) ([]*proto.UpstreamInventoryHello, error) {
+	handles, ok := a.inventory.GetAllControlStreams(hostID)
+	if !ok {
+		return nil, trace.NotFound("no control streams for host %v", hostID)
+	}
+	now := a.clock.Now()
+
+	qualifiedHellos := make([]*proto.UpstreamInventoryHello, 0)
```

Comment:
```suggestion
	var qualifiedHellos []*proto.UpstreamInventoryHello
```
https://go.dev/wiki/CodeReviewComments#declaring-empty-slices

(Or add a comment if you're avoiding nil slice intentionally for some reason.)